File: 39-combination-sum/39.py
class Solution:
    def find(self, nums, index, target):
        if sum(self.sets) == target:
            self.ans.append(self.sets.copy())
            return
        elif sum(self.sets) > target:
            return
        elif index >= len(nums):
            return

        self.find(nums, index+1, target)

        self.sets.append(nums[index])
        self.find(nums, index, target)
        self.sets.pop()

    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        self.sets = []
        self.ans = []
        self.find(candidates, 0, target)

        return self.ans

File: 39-combination-sum/39.cpp
class Solution {
private:
    vector<int> tmp;
    vector<int> can;
    vector<vector<int>> ans;
    void bt(int index, int target) {
        if (target == 0) {
            ans.push_back(tmp);
            return;
        }

        for (int i = index; i < can.size(); ++i) {
            int n = can[i];
            if (n <= target) {
                tmp.push_back(n);
                bt(i, target - n);
                tmp.pop_back();
            }
        }
    }
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        can = candidates;
        bt(0, target);

        return ans;
    }
};
